### PR TITLE
Added field CollisionCount to StatefulSetStatus

### PR DESCRIPTION
**Pull request by liyinan926**
**What this PR does / why we need it**:
This PR added a new field `CollisionCount` into `StatefulSetStatus`, similarly in terms of both name and semantics to the existing `CollisionCount` field in `DaemonSetStatus`.  The field will be used for collision avoidance when the `StatefulSet` controller creates name for the newest ControllerRevision, which will be done in another PR.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: #49909.

**Special notes for your reviewer**:
A second PR will include logic that actually uses the field for collision avoidance.

**Release note**:
```release-note
Added field CollisionCount to StatefulSetStatus in both apps/v1beta1 and apps/v1beta2
```

### CODE REVIEW

**Comment by liyinan926** [comment]
CLA signed.

File: pkg/apis/apps/types.go [hunk 1]
```diff
@@ -187,6 +187,12 @@ type StatefulSetStatus struct {
 	// updateRevision, if not empty, indicates the version of the StatefulSet used to generate Pods in the sequence
 	// [replicas-updatedReplicas,replicas)
 	UpdateRevision string
+
+	// Count of hash collisions for the StatefulSet. The StatefulSet controller
```
Comment (supereagle) [inline pkg/apis/apps/types.go]: Please keep the code style, like "collisionCount is the count of hash collisions for the StatefulSet ..."
Please also fix the code style in other places for this field.

Comment (liyinan926) [inline pkg/apis/apps/types.go]: Done.

**Comment by liyinan926** [comment]
@caesarxuchao PTAL. Thanks!

File: pkg/apis/apps/validation/validation_test.go [hunk 2]
```diff
@@ -20,8 +20,11 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/davecgh/go-spew/spew"
```
Comment (kow3ns) [inline pkg/apis/apps/validation/validation_test.go]: Why are we importing spew?

File: pkg/apis/apps/validation/validation_test.go [hunk 3]
```diff
@@ -300,6 +303,120 @@ func TestValidateStatefulSet(t *testing.T) {
...
+			ReadyReplicas:      test.readyReplicas,
+			CurrentReplicas:    test.currentReplicas,
+			UpdatedReplicas:    test.updatedReplicas,
+			ObservedGeneration: test.observedGeneration,
+			CollisionCount:     test.collisionCount,
+		}
+
+		errs := ValidateStatefulSetStatus(&status, field.NewPath("status"))
+		if hasErr := len(errs) > 0; hasErr != test.expectedErr {
+			errString := spew.Sprintf("%#v", errs)
```
Comment (kow3ns) [inline pkg/apis/apps/validation/validation_test.go]: You don't need to do this, there's a utility to aggregate errors into a single error in api machinery.

**Review by kow3ns — COMMENTED** [review]
Generally LGTM. Please see comment about the use of spew directly. Also, please update the release note to indicate that both versions GroupVersions of the StatefulSet API are affected.

File: pkg/apis/apps/validation/validation_test.go (hunk 3 above)
Comment (liyinan926) [inline pkg/apis/apps/validation/validation_test.go]: Done.

**Comment by liyinan926** [comment]
@kow3ns updated the release note. 
/retest

**Comment by liyinan926** [comment]
@kow3ns All Jenkins jobs succeeded except for `pull-kubernetes-e2e-gce-etcd3` that failed on 1 test case `Garbage collector [It] should not be blocked by dependency circle`, which was reported to be flaky in #50047. PTAL. Thanks!

**Comment by liyinan926** [comment]
@kow3ns Jenkins jobs are all good now.

**Comment by kow3ns** [comment]
you need an api reviewer

**Comment by liyinan926** [comment]
@thockin Need your approval. PTAL. Thanks!

File: pkg/apis/apps/types.go [hunk 4]
```diff
@@ -187,6 +187,12 @@ type StatefulSetStatus struct {
 	// updateRevision, if not empty, indicates the version of the StatefulSet used to generate Pods in the sequence
 	// [replicas-updatedReplicas,replicas)
 	UpdateRevision string
+
+	// collisionCount is the count of hash collisions for the StatefulSet. The StatefulSet controller
+	// uses this field as a collision avoidance mechanism when it needs to create the name for the
+	// newest ControllerRevision.
+	// +optional
+	CollisionCount *int64
```
Comment (thockin) [inline pkg/apis/apps/types.go]: Why is this int64 and not int32?  Do you expect more than 2 billion collisions?  JSON is known to have trouble carrying int64 values, which can make automated API fuzzing difficult.

Comment (liyinan926) [inline pkg/apis/apps/types.go]: Personally I don't think int64 is necessary, but it's the type for the field in other controller objects, e.g., https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/types.go#L345 and https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/types.go#L519. We want the field to have consistent type across controllers.

**Comment by liyinan926** [comment]
@thockin replied to your comment. PTAL. Thanks!

**Comment by liyinan926** [comment]
@kow3ns need lgtm again after rebasing. @thockin need you approval. PTAL. Thanks!

File: pkg/apis/apps/types.go (hunk 4 above)
Comment (thockin) [inline pkg/apis/apps/types.go]: Consistent with wrong is still wrong.  I'd like to hear a reason this can't be int32, and the others fixed.

Comment (liyinan926) [inline pkg/apis/apps/types.go]: Will update to use int32 and have a follow up PR to change it to int32 in other controllers.

Comment (kow3ns) [inline pkg/apis/apps/types.go]: @thockin 
> JSON is known to have trouble carrying int64 values, which can make automated API fuzzing difficult.

This is not an issue that users have actually reported having, and fuzzing currently works for these API objects. If you configure your decoder appropriately or use structs, or classes depending on the language, to decode, disambiguating int64 values is definitely achievable without much work.

Given that the API for [Deployment](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/deployment.md) and [DeamonSet](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md) both went through community design review prior to implementation and merge, that there is no measurable performance benefit to using an int32 in this case, that we use int64 in many other places in the API, and that we don't have an actual issue to motivate changing all the workloads API objects, I don't see why we should switch this field to int32.

Comment (liyinan926) [inline pkg/apis/apps/types.go]: @thockin What are your thoughts on the comment @kow3ns made above?

Comment (thockin) [inline pkg/apis/apps/types.go]: Discussed.  We should convert to int32 in bulk.

Comment (liyinan926) [inline pkg/apis/apps/types.go]: SGTM. Will have a PR to convert all.

**Comment by caesarxuchao** [comment]
@liyinan926 do you remember why the PR had two merge commits "Merge branch 'master' of github.com:kubernetes/kubernetes"? Did you merge your local branch with kubernetes/kubernetes master branch before submitting the PR?

@sttts reported this PR broke the publishing robot because the robot skipped merge commits while in case of this PR, the merge commits are non-empty.

**Comment by liyinan926** [comment]
@caesarxuchao @sttts I don't remember exactly why but I did pulled the master branch and rebased before the PR was approved and submitted.

**Pull request merged**